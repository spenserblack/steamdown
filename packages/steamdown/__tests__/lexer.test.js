const { lexAllInline, inlineTokens, inlineLexer } = require("../src/lexer");

describe("lexer", () => {
  describe.each([
    ["Hello, World!", [[inlineTokens.text, "Hello, World!", { x: 0, y: 0 }]]],
    [
      "one\nparagraph\nspanning lines",
      [
        [inlineTokens.text, "one", { x: 0, y: 0 }],
        [inlineTokens.newline, "\n", { x: 3, y: 0 }],
        [inlineTokens.text, "paragraph", { x: 0, y: 1 }],
        [inlineTokens.newline, "\n", { x: 9, y: 1 }],
        [inlineTokens.text, "spanning lines", { x: 0, y: 2 }],
      ],
    ],
    [
      "*Italicized text*",
      [
        [inlineTokens.asterisk, "*", { x: 0, y: 0 }],
        [inlineTokens.text, "Italicized text", { x: 1, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 16, y: 0 }],
      ],
    ],
    [
      "Nested *italicized* text",
      [
        [inlineTokens.text, "Nested ", { x: 0, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 7, y: 0 }],
        [inlineTokens.text, "italicized", { x: 8, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 18, y: 0 }],
        [inlineTokens.text, " text", { x: 19, y: 0 }],
      ],
    ],
    [
      "*Italics* and *more italics*",
      [
        [inlineTokens.asterisk, "*", { x: 0, y: 0 }],
        [inlineTokens.text, "Italics", { x: 1, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 8, y: 0 }],
        [inlineTokens.text, " and ", { x: 9, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 14, y: 0 }],
        [inlineTokens.text, "more italics", { x: 15, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 27, y: 0 }],
      ],
    ],
    [
      "_Underlined text_",
      [
        [inlineTokens.underscore, "_", { x: 0, y: 0 }],
        [inlineTokens.text, "Underlined text", { x: 1, y: 0 }],
        [inlineTokens.underscore, "_", { x: 16, y: 0 }],
      ],
    ],
    [
      "*Italicized and _underlined_*",
      [
        [inlineTokens.asterisk, "*", { x: 0, y: 0 }],
        [inlineTokens.text, "Italicized and ", { x: 1, y: 0 }],
        [inlineTokens.underscore, "_", { x: 16, y: 0 }],
        [inlineTokens.text, "underlined", { x: 17, y: 0 }],
        [inlineTokens.underscore, "_", { x: 27, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 28, y: 0 }],
      ],
    ],
    [
      "**Bold text**",
      [
        [inlineTokens.asterisk, "*", { x: 0, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 1, y: 0 }],
        [inlineTokens.text, "Bold text", { x: 2, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 11, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 12, y: 0 }],
      ],
    ],
    [
      "Nested **bold** text",
      [
        [inlineTokens.text, "Nested ", { x: 0, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 7, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 8, y: 0 }],
        [inlineTokens.text, "bold", { x: 9, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 13, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 14, y: 0 }],
        [inlineTokens.text, " text", { x: 15, y: 0 }],
      ],
    ],
    [
      "***Bold and italicized text***",
      [
        [inlineTokens.asterisk, "*", { x: 0, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 1, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 2, y: 0 }],
        [inlineTokens.text, "Bold and italicized text", { x: 3, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 27, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 28, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 29, y: 0 }],
      ],
    ],
    [
      "~strike text~",
      [
        [inlineTokens.tilde, "~", { x: 0, y: 0 }],
        [inlineTokens.text, "strike text", { x: 1, y: 0 }],
        [inlineTokens.tilde, "~", { x: 12, y: 0 }],
      ],
    ],
    [
      "~~strike text~~",
      [
        [inlineTokens.tilde, "~", { x: 0, y: 0 }],
        [inlineTokens.tilde, "~", { x: 1, y: 0 }],
        [inlineTokens.text, "strike text", { x: 2, y: 0 }],
        [inlineTokens.tilde, "~", { x: 13, y: 0 }],
        [inlineTokens.tilde, "~", { x: 14, y: 0 }],
      ],
    ],
    [
      "\\*escaped\\*",
      [
        [inlineTokens.escape, "\\", { x: 0, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 1, y: 0 }],
        [inlineTokens.text, "escaped", { x: 2, y: 0 }],
        [inlineTokens.escape, "\\", { x: 9, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 10, y: 0 }],
      ],
    ],
    [
      "*italics \\*escaped\\**",
      [
        [inlineTokens.asterisk, "*", { x: 0, y: 0 }],
        [inlineTokens.text, "italics ", { x: 1, y: 0 }],
        [inlineTokens.escape, "\\", { x: 9, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 10, y: 0 }],
        [inlineTokens.text, "escaped", { x: 11, y: 0 }],
        [inlineTokens.escape, "\\", { x: 18, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 19, y: 0 }],
        [inlineTokens.asterisk, "*", { x: 20, y: 0 }],
      ],
    ],
    [
      "Use a \\\\ to escape special characters",
      [
        [inlineTokens.text, "Use a ", { x: 0, y: 0 }],
        [inlineTokens.escape, "\\", { x: 6, y: 0 }],
        [inlineTokens.escape, "\\", { x: 7, y: 0 }],
        [inlineTokens.text, " to escape special characters", { x: 8, y: 0 }],
      ],
    ],
    [
      "{not parsed}",
      [
        [inlineTokens.leftCurly, "{", { x: 0, y: 0 }],
        [inlineTokens.text, "not parsed", { x: 1, y: 0 }],
        [inlineTokens.rightCurly, "}", { x: 11, y: 0 }],
      ],
    ],
  ])("inlineLexer(`%s`)", (text, expected) => {
    test("lexes all inline tokens", () => {
      const [next, hasNext] = inlineLexer(text);
      expected.forEach((lexed) => {
        expect(hasNext()).toBe(true);
        expect(next()).toEqual(lexed);
      });
      expect(hasNext()).toBe(false);
    });

    test("snapshot", () => {
      expect(lexAllInline(text)).toMatchSnapshot();
    });
  });
});
