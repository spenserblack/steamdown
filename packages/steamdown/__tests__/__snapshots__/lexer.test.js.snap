// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer inlineLexer(\`***Bold and italicized text***\`) snapshot 1`] = `
[
  [
    Symbol(*),
    "*",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 2,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "Bold and italicized text",
    {
      "x": 3,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 27,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 28,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 29,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`**Bold text**\`) snapshot 1`] = `
[
  [
    Symbol(*),
    "*",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "Bold text",
    {
      "x": 2,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 11,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 12,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`*Italicized and _underlined_*\`) snapshot 1`] = `
[
  [
    Symbol(*),
    "*",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "Italicized and ",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(_),
    "_",
    {
      "x": 16,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "underlined",
    {
      "x": 17,
      "y": 0,
    },
  ],
  [
    Symbol(_),
    "_",
    {
      "x": 27,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 28,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`*Italicized text*\`) snapshot 1`] = `
[
  [
    Symbol(*),
    "*",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "Italicized text",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 16,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`*Italics* and *more italics*\`) snapshot 1`] = `
[
  [
    Symbol(*),
    "*",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "Italics",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 8,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    " and ",
    {
      "x": 9,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 14,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "more italics",
    {
      "x": 15,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 27,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`*italics \\*escaped\\**\`) snapshot 1`] = `
[
  [
    Symbol(*),
    "*",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "italics ",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(\\),
    "\\",
    {
      "x": 9,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 10,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "escaped",
    {
      "x": 11,
      "y": 0,
    },
  ],
  [
    Symbol(\\),
    "\\",
    {
      "x": 18,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 19,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 20,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`>!spoiler text!<\`) snapshot 1`] = `
[
  [
    Symbol(>),
    ">",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(!),
    "!",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "spoiler text",
    {
      "x": 2,
      "y": 0,
    },
  ],
  [
    Symbol(!),
    "!",
    {
      "x": 14,
      "y": 0,
    },
  ],
  [
    Symbol(<),
    "<",
    {
      "x": 15,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`\\*escaped\\*\`) snapshot 1`] = `
[
  [
    Symbol(\\),
    "\\",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "escaped",
    {
      "x": 2,
      "y": 0,
    },
  ],
  [
    Symbol(\\),
    "\\",
    {
      "x": 9,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 10,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`_Underlined text_\`) snapshot 1`] = `
[
  [
    Symbol(_),
    "_",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "Underlined text",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(_),
    "_",
    {
      "x": 16,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`{not parsed}\`) snapshot 1`] = `
[
  [
    Symbol({),
    "{",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "not parsed",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(}),
    "}",
    {
      "x": 11,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`~~strike text~~\`) snapshot 1`] = `
[
  [
    Symbol(~),
    "~",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(~),
    "~",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "strike text",
    {
      "x": 2,
      "y": 0,
    },
  ],
  [
    Symbol(~),
    "~",
    {
      "x": 13,
      "y": 0,
    },
  ],
  [
    Symbol(~),
    "~",
    {
      "x": 14,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`~strike text~\`) snapshot 1`] = `
[
  [
    Symbol(~),
    "~",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "strike text",
    {
      "x": 1,
      "y": 0,
    },
  ],
  [
    Symbol(~),
    "~",
    {
      "x": 12,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`Hello, World!\`) snapshot 1`] = `
[
  [
    Symbol(text),
    "Hello, World",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(!),
    "!",
    {
      "x": 12,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`Nested **bold** text\`) snapshot 1`] = `
[
  [
    Symbol(text),
    "Nested ",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 7,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 8,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "bold",
    {
      "x": 9,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 13,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 14,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    " text",
    {
      "x": 15,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`Nested *italicized* text\`) snapshot 1`] = `
[
  [
    Symbol(text),
    "Nested ",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 7,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "italicized",
    {
      "x": 8,
      "y": 0,
    },
  ],
  [
    Symbol(*),
    "*",
    {
      "x": 18,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    " text",
    {
      "x": 19,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`Use a \\\\ to escape special characters\`) snapshot 1`] = `
[
  [
    Symbol(text),
    "Use a ",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(\\),
    "\\",
    {
      "x": 6,
      "y": 0,
    },
  ],
  [
    Symbol(\\),
    "\\",
    {
      "x": 7,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    " to escape special characters",
    {
      "x": 8,
      "y": 0,
    },
  ],
]
`;

exports[`lexer inlineLexer(\`one
paragraph
spanning lines\`) snapshot 1`] = `
[
  [
    Symbol(text),
    "one",
    {
      "x": 0,
      "y": 0,
    },
  ],
  [
    Symbol(
),
    "
",
    {
      "x": 3,
      "y": 0,
    },
  ],
  [
    Symbol(text),
    "paragraph",
    {
      "x": 0,
      "y": 1,
    },
  ],
  [
    Symbol(
),
    "
",
    {
      "x": 9,
      "y": 1,
    },
  ],
  [
    Symbol(text),
    "spanning lines",
    {
      "x": 0,
      "y": 2,
    },
  ],
]
`;
